common/meparser.cpp

```diff
@@ -103,88 +103,135 @@ USTATUS MeParser::parseMeRegionBody(const UModelIndex & index)
+            return U_INVALID_ME_PARTITION_TABLE;
+    }
+
+    // XXX: Shall we continue with parsing?
+    if (msgInvalidPtHeaderChecksum) {
+        msg(usprintf("%s: FPT header checksum is invalid", __FUNCTION__), index);
+        return U_INVALID_ME_PARTITION_TABLE;
+    }
+
     // Get info
-    UByteArray header = region.left(romBypassVectorSize + sizeof(FPT_HEADER));
+    UByteArray header = region.left(pt1Header->HeaderLength);
     UByteArray body = region.mid(header.size(), ptBodySize);
-    
+
     UString name = UString("FPT partition table");
-    UString info = usprintf("Full size: %Xh (%u)\nHeader size: %Xh (%u)\nBody size: %Xh (%u)\nNumber of entries: %u\nHeader version: %02Xh\nEntry version: %02Xh\n"
-                            "Header length: %02Xh\nTicks to add: %04Xh\nTokens to add: %04Xh\nUMA size: %Xh\nFlash layout: %Xh\nFITC version: %u.%u.%u.%u\nChecksum: %02Xh, ",
-                            ptSize, ptSize,
-                            header.size(), header.size(),
-                            ptBodySize, ptBodySize,
-                            ptHeader->NumEntries,
-                            ptHeader->HeaderVersion,
-                            ptHeader->EntryVersion,
-                            ptHeader->HeaderLength,
-                            ptHeader->TicksToAdd,
-                            ptHeader->TokensToAdd,
-                            ptHeader->UmaSize,
-                            ptHeader->FlashLayout,
-                            ptHeader->FitcMajor, ptHeader->FitcMinor, ptHeader->FitcHotfix, ptHeader->FitcBuild,
-                            ptHeader->Checksum) + (ptHeader->Checksum == calculated ? UString("valid") : usprintf("invalid, should be %02Xh", calculated));
-    
+    UString info;
+
+    switch (pt1Header->HeaderVersion) {
+        case 0x10:
+            /* FALLTHROUGH */
+        case 0x20:
+            info = usprintf("Full size: %Xh (%u)\nHeader size: %Xh (%u)\nBody size: %Xh (%u)\nNumber of entries: %u\n"
+                            "Header version: %02Xh\nEntry version: %02Xh\nHeader length: %02Xh\n"
+                            "Ticks to add: %04Xh\nTokens to add: %04Xh\nUMA size: %Xh\nFlash layout: %Xh\n"
+                            "FITC version: %u.%u.%u.%u\nChecksum: %02Xh, ",
+                       ptSize, ptSize, header.size(), header.size(), ptBodySize, ptBodySize, pt1Header->NumEntries,
+                       pt1Header->HeaderVersion, pt1Header->EntryVersion, pt1Header->HeaderLength,
+                       pt1Header->TicksToAdd, pt1Header->TokensToAdd, pt1Header->UmaSize, pt1Header->FlashLayout,
+                       pt1Header->FitcMajor, pt1Header->FitcMinor, pt1Header->FitcHotfix, pt1Header->FitcBuild, pt1Header->Checksum);
+            info += (msgInvalidPtHeaderChecksum ? usprintf("invalid, should be %02Xh", chksumCalculated) : UString("valid"));
+            break;
+        case 0x21:
```

Comment:
Dead code due to previous checks now. 
info += (msgInvalidPtHeaderChecksum ? usprintf("invalid, should be %08Xh", chksumCalculated) : UString("valid")); line is a lie because there's no code that calculates chksumCalculated for 0x21 header type.